components/proxy/src/test/java/com/hotels/styx/admin/handlers/ServiceProviderHandlerTest.java

```diff
@@ -0,0 +1,168 @@
+        ServiceProviderHandler handler = new ServiceProviderHandler(store);
+        HttpRequest request = HttpRequest.get("/admin/service/provider/nonexistent").build();
+        HttpResponse response = Mono.from(handler.handle(request, HttpInterceptorContext.create())).block();
+
+        assertThat(response.status(), equalTo(HttpResponseStatus.NOT_FOUND));
+    }
+
+    @Test
+    public void returnsNotFoundStatusWithNonHandledUrl() {
+        StyxObjectStore<ProviderObjectRecord> empty = new StyxObjectStore<>();
+        ServiceProviderHandler handler = new ServiceProviderHandler(empty);
+        HttpRequest request = HttpRequest.get("/not/my/url").build();
+        HttpResponse response = Mono.from(handler.handle(request, HttpInterceptorContext.create())).block();
+
+        assertThat(response.status(), equalTo(HttpResponseStatus.NOT_FOUND));
+    }
+
+    private StyxObjectStore<ProviderObjectRecord> createTestStore() throws IOException {
+        ObjectMapper mapper =new ObjectMapper();
+        StyxService mockService = Mockito.mock(StyxService.class);
+        StyxObjectStore<ProviderObjectRecord> store = new StyxObjectStore<>();
+
+        JsonNode config1 = mapper.readTree("{\"setting1\" : \"A\", \"setting2\" : \"A\"}");
+        ProviderObjectRecord rec1 = new ProviderObjectRecord("Type1", new HashSet<>(Arrays.asList("this=that", "truth=false")),
+                config1, mockService);
+        store.insert("object1", rec1);
+
+        JsonNode config2 = mapper.readTree("{\"setting1\" : \"B\", \"setting2\" : \"B\"}");
+        ProviderObjectRecord rec2 = new ProviderObjectRecord("Type2", new HashSet<>(Arrays.asList("up=down", "weakness=strength")),
+                config2, mockService);
+        store.insert("object2", rec2);
+
+        JsonNode config3 = mapper.readTree("{\"setting1\" : \"C\", \"setting2\" : \"C\"}");
+        ProviderObjectRecord rec3 = new ProviderObjectRecord("Type3", new HashSet<>(Arrays.asList("black=white", "left=right")),
+                config3, mockService);
+        store.insert("object3", rec3);
+
+        return store;
+    }
+
+    private StyxObjectDefinition deserialiseProvider(String yaml) {
+        try {
+            return ServiceProviderHandler.yamlMapper().readValue(yaml, StyxObjectDefinition.class);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private List<StyxObjectDefinition> extractProviders(String responseBody) throws IOException {
+        return Arrays.stream(responseBody.split("---\n"))
+                .filter(StringUtils::isNotBlank)
+                .map(this::deserialiseProvider)
+                .collect(Collectors.toList());
+    }
+
+    private void validateProvider(StyxObjectDefinition actual, ProviderObjectRecord expected) {
+        assertThat(actual.type(), equalTo(expected.getType()));
+        assertThat(actual.tags(), containsInAnyOrder(expected.getTags().toArray()));
```

Comment:
I would also check that it doesn't contain any other elements (e.g., verify the length).

Reply:
I think `containsInAnyOrder` also requires the length to be equal.

Reply:
Oh right. The name always gets me confused.